### PR TITLE
Constructable & Linkable Shutters/Blastdoor + Mini-Refactor

### PR DESCRIPTION
**Pull request by redmoogle**
### Intent of your Pull Request

Add craftable Shutters and Blastdoors

<details>
  <summary>Shutter Crafting</summary>

-  Under Misc Category
- 10 Plasteel
- 5 Cables
- 1 Airlock Electronics
- Tools: Welder, Screwdriver, Wirecutters
- Time: 10 Seconds

</details>
<details>
  <summary>Blastdoor Crafting</summary>

-  Under Misc Category
- 20 Plasteel
- 10 Cables
- 1 Airlock Electronics
- Tools: Welder, Screwdriver, Wirecutters
- Time: 20 Seconds

</details>
<details>
  <summary>Linking/Unlinking Process</summary>

### Linking
- Get a blast door controller from the autolathe
- Craft a button frame with metal
- Put door controller into button frame(probably should put a airlock controller for access too)
- Use multitool to generate Unique ID
- Use Screwdriver to open the panel of the door
- Click the opened panel with the multitool
- Your door is now linked

### Unlinking
- Open the panel
- Use wirecutters
- Wait 10 seconds

</details>
<details>
  <summary>Deconstruction</summary>

- Have Door Open
- Use Welder and wait 10 seconds
- Use Crowbar and wait 15 seconds

</details>

### Why is this good for the game?

Allows more items to be crafted and used. _Might be part of my plan to make everything craftable ~~wait what :thonk:~~_

#### Changelog

:cl:  
rscadd: Add the ability to link blastdoors/shutters
rscadd: Add the ability to create blastdoors/shutters
/:cl:

### CODE REVIEW

**Comment by patpol4** [comment]
hot

**Comment by Xoxeyos** [comment]
Will you add a way to deconstruct these?

**Comment by redmoogle** [comment]
> 
> 
> Will you add a way to deconstruct these?

yes

**Comment by maxion12345** [comment]
ew crafting menu crafting

while your add it add a way to deconstruct toilets

**Comment by redmoogle** [comment]
> 
> 
> ew crafting menu crafting
> 
> while your add it add a way to deconstruct toilets

Crafting Menu is good for when your lazy, tbh it should be intergrated into the z-metal/mat menu

also atomization

**Comment by MsSeraph** [comment]
Absolute chad, and if anyone says "OMG HOW WILL YOU GET THROUGH THIS ITS SO OP", you can depower them to crowbar them open, you can EMP them to depower them and you can use an APC control console to depower them too, so you still can get through, its not impenetrable, thus not OP, good PR

File: code/game/machinery/buttons.dm [hunk 1]
```diff
@@ -89,12 +89,32 @@
...
+				var/obj/item/multitool/P = W
+				if(!id) // Generate New ID if none exists
+					id = rand(1, 25565) // rare enough that ids should never conflict
+					to_chat(user, "<span class='notice'>No ID found. Generating New ID</span>")
+
+				P.buffer = id
+				to_chat(user, "<span class='notice'>You link the button to the [P].</span>")
+				setup_device() // Has to be done. It sets the signaller up
+			else
+				to_chat(user, "[device ? "" : "<span class='notice'>No blast door controller found</span>"]")
```
Comment (Bibby0110) [inline code/game/machinery/buttons.dm]: why not just an if statement?

File: code/game/machinery/buttons.dm [hunk 2]
```diff
@@ -89,12 +89,32 @@
...
+
+		if(board && W.tool_behaviour == TOOL_WIRECUTTER)
+			to_chat(user, "<span class='notice'>You start wiping the button ID...</span>")
```
Comment (Bibby0110) [inline code/game/machinery/buttons.dm]:
```suggestion
			to_chat(user, "<span class='notice'>You start wiping the button's ID...</span>")
```

File: code/game/machinery/buttons.dm [hunk 3]
```diff
@@ -89,12 +89,32 @@
...
+			W.play_tool_sound(src)
+			if(W.use_tool(src, user, 4 SECONDS)) // To make sure they're not doing this unintentionally
+				to_chat(user, "<span class='notice'>You wipe the buttons ID.</span>")
```
Comment (Bibby0110) [inline code/game/machinery/buttons.dm]:
```suggestion
				to_chat(user, "<span class='notice'>You wipe the button's ID.</span>")
```

File: code/game/machinery/doors/poddoor.dm [hunk 4]
```diff
@@ -96,3 +101,85 @@
...
+
+/obj/machinery/door/poddoor/attackby(obj/item/W, mob/user, params)
+	if(special) // No Cheesing
+		to_chat(user, "<span class='warning'>This door appears to have a different screw.</span>")
+		return
+
+
+	if(W.tool_behaviour == TOOL_SCREWDRIVER)
+		if(density)
+			to_chat(user, "<span class='warning'>You need to open [src] before opening its maintenence panel.</span>")
```
Comment (Bibby0110) [inline code/game/machinery/doors/poddoor.dm]:
```suggestion
			to_chat(user, "<span class='warning'>You need to open [src] before opening its maintenance panel.</span>")
```

File: code/game/machinery/doors/poddoor.dm [hunk 5]
```diff
@@ -15,6 +15,8 @@
 	resistance_flags = FIRE_PROOF
 	damage_deflection = 70
 	poddoor = TRUE
+	var/special = FALSE // Prevents ERT or whatever from breaking into their shutters
+	var/constructionstate = "intact" // Decounstruction Stuff
```
Comment (Bibby0110) [inline code/game/machinery/doors/poddoor.dm]: Use defines instead of strings

File: code/game/machinery/doors/poddoor.dm [hunk 6]
```diff
@@ -96,3 +101,85 @@
...
+			if(P.amount < 5)
+				to_chat(user, "<span class='warning'>You need 5 plasteel sheets to put the plating back on</span>")
+				return
+			
+			P.use(5)
+			constructionstate = "intact"
+			return
+
+	else
+		to_chat(user, "<span class='warning'>You need to open the panel.</span>")
```
Comment (Bibby0110) [inline code/game/machinery/doors/poddoor.dm]:
```suggestion
		to_chat(user, "<span class='warning'>You need to open the panel first.</span>")
```

File: code/game/machinery/doors/poddoor.dm [hunk 7]
```diff
@@ -96,3 +101,85 @@
...
+					new /obj/item/stack/sheet/plasteel(loc, 15)
+					new /obj/item/electronics/airlock(loc)
+					new /obj/item/stack/cable_coil/red(loc, 10)
+
+				qdel(src)
+
+		if(istype(W, /obj/item/stack/sheet/plasteel))
+			var/obj/item/stack/sheet/plasteel/P = W
+			if(P.amount < 5)
+				to_chat(user, "<span class='warning'>You need 5 plasteel sheets to put the plating back on</span>")
```
Comment (Bibby0110) [inline code/game/machinery/doors/poddoor.dm]:
```suggestion
				to_chat(user, "<span class='warning'>You need 5 plasteel sheets to put the plating back on.</span>")
```

File: code/game/machinery/doors/poddoor.dm [hunk 8]
```diff
@@ -96,3 +101,85 @@
...
+				else
+					new /obj/item/stack/sheet/plasteel(loc, 15)
+					new /obj/item/electronics/airlock(loc)
+					new /obj/item/stack/cable_coil/red(loc, 10)
+
+				qdel(src)
+
+		if(istype(W, /obj/item/stack/sheet/plasteel))
+			var/obj/item/stack/sheet/plasteel/P = W
+			if(P.amount < 5)
```
Comment (Bibby0110) [inline code/game/machinery/doors/poddoor.dm]:
```suggestion
			if(P.use(5))
```

File: code/game/machinery/doors/poddoor.dm [hunk 9]
```diff
@@ -96,3 +101,85 @@
...
+				to_chat(user, "<span class='warning'>You need 5 plasteel sheets to put the plating back on</span>")
+				return
+			
+			P.use(5)
```
Comment (Bibby0110) [inline code/game/machinery/doors/poddoor.dm]:
```suggestion
```

File: code/game/machinery/doors/poddoor.dm [hunk 10]
```diff
@@ -96,3 +101,85 @@
...
+			playsound(src.loc, 'sound/items/welder.ogg', 50, 1)
+			if(do_after(user, 10 SECONDS, target = src))
+				to_chat(user, "<span class='notice'>You remove the outer plasteel cover.</span>")
+				constructionstate = "welded"
+				id = null // Effectivley breaks the door
+				new /obj/item/stack/sheet/plasteel(loc, 5)
+				return
+		
+		if(W.tool_behaviour == TOOL_CROWBAR && constructionstate == "welded")
+			to_chat(user, "<span class='notice'>You start to remove all of the internal components</span>")
```
Comment (Bibby0110) [inline code/game/machinery/doors/poddoor.dm:157]:
```suggestion
			to_chat(user, "<span class='notice'>You start to remove all of the internal components.</span>")
```

File: code/game/machinery/doors/poddoor.dm [hunk 11]
```diff
@@ -96,3 +101,85 @@
...
+		if(density)
+			to_chat(user, "<span class='warning'>You need to open [src] before opening its maintenence panel.</span>")
+			return
+		else if(default_deconstruction_screwdriver(user, icon_state, icon_state, W))
+			to_chat(user, "<span class='notice'>You [panel_open ? "open" : "close"] the maintenance hatch of [src].</span>")
+			return TRUE
+
+	if(panel_open)
+		if(W.tool_behaviour == TOOL_MULTITOOL && constructionstate == "intact")
+			if(id != null)
```
Comment (Bibby0110) [inline code/game/machinery/doors/poddoor.dm:121]:
```suggestion
			if(id)
```

**Review by Bibby0110 — CHANGES_REQUESTED** [review]
fine code

File: code/game/machinery/buttons.dm (hunk 1 above)
Comment (redmoogle) [reply code/game/machinery/buttons.dm]: er what. it just tells you if you dont have a blast door controller in it

Comment (Bibby0110) [reply code/game/machinery/buttons.dm]: yeah but then why do you have the device ? statement, either remove that or add an if statement before the to_chat
(preferably the first option)

**Review by fluffe9911 — APPROVED** [review]
the id generation feels pretty snowflakey but seems fine overall

**Pull request merged**